flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionService.java

```diff
@@ -0,0 +1,219 @@
+		this.leaderContender = null;
+		this.configMapLabels = KubernetesUtils.getConfigMapLabels(
+			leaderConfig.getClusterId(), LABEL_CONFIGMAP_TYPE_HIGH_AVAILABILITY);
+	}
+
+	@Override
+	public void internalStart(LeaderContender contender) {
+		CompletableFuture.runAsync(leaderElector::run, executor);
+		kubernetesWatch = kubeClient.watchConfigMaps(configMapName, new ConfigMapCallbackHandlerImpl());
+	}
+
+	@Override
+	public void internalStop() {
+		if (kubernetesWatch != null) {
+			kubernetesWatch.close();
+		}
+	}
+
+	@Override
+	protected void writeLeaderInformation() {
+		try {
+			kubeClient.checkAndUpdateConfigMap(
+				configMapName,
+				configMap -> {
+					if (leaderElector.hasLeadership(configMap)) {
+						// Get the updated ConfigMap with new leader information
+						if (confirmedLeaderAddress != null && confirmedLeaderSessionID != null) {
+							configMap.getData().put(LEADER_ADDRESS_KEY, confirmedLeaderAddress);
+							configMap.getData().put(LEADER_SESSION_ID_KEY, confirmedLeaderSessionID.toString());
+						}
+						configMap.getLabels().putAll(configMapLabels);
+						return Optional.of(configMap);
+					}
+					return Optional.empty();
+				}).get();
+		} catch (Exception e) {
+			leaderContender.handleError(new Exception("Could not update ConfigMap " + configMapName, e));
+		}
+	}
+
+	@Override
+	protected boolean hasLeadership() {
+		return kubeClient.getConfigMap(configMapName)
+			.map(leaderElector::hasLeadership)
+			.orElse(false);
+	}
+
+	@Override
+	public String toString() {
+		return "KubernetesLeaderElectionService{configMapName='" + configMapName + "'}";
+	}
+
+	private class LeaderCallbackHandlerImpl extends KubernetesLeaderElector.LeaderCallbackHandler {
+
+		@Override
+		public void isLeader() {
+			onGrantLeadership();
+		}
+
+		@Override
+		public void notLeader() {
+			// Clear the leader information in ConfigMap
+			try {
+				kubeClient.checkAndUpdateConfigMap(
+					configMapName,
+					configMap -> {
+						// Do not need to check the leader here
+						configMap.getData().remove(LEADER_ADDRESS_KEY);
+						configMap.getData().remove(LEADER_SESSION_ID_KEY);
+						return Optional.of(configMap);
+					}
+				).get();
+			} catch (Exception e) {
+				leaderContender.handleError(
+					new Exception("Could not remove leader information from ConfigMap " + configMapName, e));
+			}
+			onRevokeLeadership();
+			// Continue to contend the leader
+			CompletableFuture.runAsync(leaderElector::run, executor);
+		}
+	}
+
+	private class ConfigMapCallbackHandlerImpl implements FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> {
+
+		@Override
+		public void onAdded(List<KubernetesConfigMap> configMaps) {
+			// noop
+		}
+
+		@Override
+		public void onModified(List<KubernetesConfigMap> configMaps) {
+			if (hasLeadership()) {
+				configMaps.forEach(configMap -> {
+					if (isLeaderUpdatedExternally(configMap)) {
+						if (configMap.getName().equals(configMapName)) {
```

Comment:
Shouldn't we check the config map name before checking whether it is updated externally?

Reply:
Why would we receive events for config map with another name? Shouldn't we only received events for the config map that we watch?
Maybe we should add an assertion for the config map name here.

Reply:
Yes. Actually, we should only receive the watched ConfigMap event. Adding a `assert` here make more senses.